Add unit tests for rolling upgrade

pebrc opened the pull request:
Add some missing unit tests for rolling sset upgrades 

Related to #1156

anyasabo commented on operators/pkg/controller/elasticsearch/driver/upgrade_test.go:
```diff
@@ -0,0 +1,321 @@
+				t.Errorf("doRollingUpgrade() = %v, want %v,err = %v, want %v", result, wantRes, err, wantErr)
+			}
+			if diff := deep.Equal(map[string]int32(mu), tt.wantNewPartition); diff != nil {
+				t.Error(diff)
```
nit: this could be `assert.Nil(t, diff)` which might make it clearer what it's checking if it fails

pebrc replied:
```
FAIL: Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset (0.00s)
        upgrade_test.go:196: 
            	Error Trace:	upgrade_test.go:196
            	Error:      	Expected nil, but got: []string{"map[data]: 0 != 3"}
            	Test:       	Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset
```
vs
```
  --- FAIL: Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset (0.00s)
        upgrade_test.go:196: [map[data]: 0 != 3]
```

pebrc replied:
Not sure which is better, second one without `assert.Nil` is more to the point but maybe lacks some context?

anyasabo replied:
Nah that makes sense to me, I retract my nit

anyasabo commented on operators/pkg/controller/elasticsearch/driver/upgrade_test.go:
```diff
@@ -0,0 +1,321 @@
+			}
+			mu := mockUpdater{}
+			fc := fakeESClient{}
+			if got := d.doRollingUpgrade(tt.args.statefulSets, &fc, tt.args.esState, mu.updatePartition); !reflect.DeepEqual(got, tt.want) {
```
My first thought here was that this could be an `assert.DeepEqual()` call, but looks like we want the information provided by `Aggregate()` to be able to debug better?

pebrc replied:
testify does not have an `assert.DeepEqual` afaik

pebrc replied:
I got rid of the `Aggregate()`

anyasabo replied:
>testify does not have an assert.DeepEqual afaik

Ah you're right, I landed on the godoc of a different assert package 🥇 . Looks like assert.Equal ends up calling reflect.DeepEqual though

anyasabo commented on operators/pkg/controller/elasticsearch/driver/upgrade_test.go:
```diff
@@ -0,0 +1,321 @@
+				statefulSets: sset.StatefulSetList{
+					nodespec.TestSset{
+						Name: "default",
+						//	Version:   "7.3.0",
```
Should this be removed?

pebrc replied:
absolutely!

barkbay commented on operators/pkg/controller/elasticsearch/driver/upgrade_test.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+func Test_defaultDriver_doRollingUpgrade(t *testing.T) {
+	// This does not test: podUpgradeDone or prepareClusterForNodeRestart in detail but is focused on the main invariants
```
Is this something to be addressed in an other PR ?
I'm 👍 with the tests, my only concern is that there is no test with a mix of "upgraded" and "yet to be upgraded" pods.

sebgl replied:
+1 on adding tests that cover some specifics of podUpgradeDone. Totally fine in another PR (and can be done by someone else).

pebrc replied:
No worries happy to spend a little more time on it

pebrc commented:
@barkbay @sebgl if you could take another look, I added the test for partially rolled out upgrades and `podUpgradeDone` 

Also, to facilitate testing, I re-introduced some of the driver indirection we removed, but, as I think, fairly well contained and restricted to the upgrade method.

sebgl commented on operators/pkg/controller/elasticsearch/driver/upgrade_test.go:
```diff
@@ -168,9 +178,35 @@ func Test_defaultDriver_doRollingUpgrade(t *testing.T) {
+				esState: defaultESState,
+			},
+			upgradedPods: map[string]bool{
+				"default-1": true, //
```
so much suspense

pebrc replied:
it's like the three dots in a chat window when the other person types ...
